packages/library/src/commonMain/kotlin/io/realm/RealmList.kt

```diff
@@ -92,34 +103,48 @@ class RealmList<E> private constructor(
                     Float::class,
                     Double::class,
                     String::class -> value
-                    else -> when {
-                        isRealmObject -> (value as Link).toRealmObject(
-                            clazz as KClass<out RealmObject>,
-                            mediator,
-                            realm
-                        )
-                        else -> throw IllegalArgumentException("Unsupported type '$clazz'.")
-                    }
+                    else -> (value as Link).toRealmObject(
+                        clazz as KClass<out RealmObject>,
+                        mediator,
+                        realm
+                    )
                 } as E
             }
         }
-
-        /**
-         * Checks if the Realm associated to this RealmList is still accessible, throwing an
-         * [IllegalStateException] if not.
-         */
-        fun checkRealmClosed() {
-            metadata.realm.checkClosed()
-        }
     }
 }
 
+/**
+ * Delegate interface for Realm-specific methods.
+ */
+internal interface RealmListApi<E> : MutableList<E> {
+
+    val listPtr: NativePointer
+
+    fun observe(list: RealmList<E>): Flow<RealmList<E>>
+    fun freeze(frozenRealm: RealmReference): RealmList<E>
+    fun thaw(liveRealm: RealmReference): RealmList<E>
+}
+
 /**
  * Represents an unmanaged [RealmList] backed by a [MutableList] via class delegation.
  */
 private class UnmanagedListDelegate<E>(
     list: MutableList<E> = mutableListOf()
-) : MutableList<E> by list
+) : MutableList<E> by list, RealmListApi<E> {
```

Comment:
Couldn't we guard about reaching this situation further out the call stack if we just checked if the list was managed or unmanaged when observing? I guess the whole observability only needs to be implemented for managed lists. 

Maybe an `Observable` interface with `observe(): Flow<T>` that we could reuse on all observables (result, object, lists) to generically dispatch to the notifier.

Reply:
Maybe we should find a way to ensure by type that it was only the managed list that was passed around in the internal implementation.

Reply:
I recall a conversation where it sounded to me like we had mixed feelings about the use of interfaces like `RealmObservable`, `RealmFreezable`, etc., that's why I didn't do it. But I 100% agree with the approach.

Reply:
Maybe we should merge this PR as is, and then discuss this abstraction while we are looking at the other interfaces?

Reply:
Yep. Think it would be easier to play around if we rearrange the interfaces in one PR. We just have to remember all of the _point of interests_.